helper/resource/testing.go

```diff
@@ -959,6 +960,34 @@ func testCheckResourceAttr(is *terraform.InstanceState, name string, key string,
 	return nil
 }
 
+// CheckResourceAttrWithFunc is the callback type used to apply a custom checking logic
+// when using TestCheckResourceAttrWith and a value is found for the given name and key.
+//
+// When this function returns an error, TestCheckResourceAttrWith will fail the check.
+type CheckResourceAttrWithFunc func(value string) error
+
+// TestCheckResourceAttrWith - as per TestCheckResourceAttr but the attribute value
+// checking logic can be customised.
+//
+// If a value is found for the given name and key, it is passed to the CheckResourceAttrWithFunc function.
+// The CheckResourceAttrWithFunc function can then apply any checking logic:
+// should return an error for the check to fail, or `nil` to succeed.
+func TestCheckResourceAttrWith(name, key string, f CheckResourceAttrWithFunc) TestCheckFunc {
+	return checkIfIndexesIntoTypeSet(key, func(s *terraform.State) error {
+		is, err := primaryInstanceState(s, name)
+		if err != nil {
+			return err
+		}
+
+		err = testCheckResourceAttrSet(is, name, key)
+		if err != nil {
+			return err
+		}
+
+		return f(is.Attributes[key])
```

Comment:
It'd be awesome to split this so the testing framework can add error context with the name and key automatically:

```suggestion
		err = f(is.Attributes[key])

		if err != nil {
			return fmt.Errorf("%s: Attribute %q value: %w", name, key, err)
		}

		return nil
```